### PR TITLE
Alerting: Improve notification policies created during migration

### LINKED ISSUE #56582: Deduplicate migrated notification policies
As a followup to #45443, I'd like to re-raise one specific item that wasn't addressed upon closing that issue, the first caveat mentioned in https://github.com/grafana/grafana/issues/45443#issuecomment-1088148604:

> I didn't tackle the matching label on notification policy using a random rule_uid. I agree that this isn't optimal, it's difficult to see at a glance which alert each policy is supposed to reference. That being said, I can't think of a simple solution that doesn't have major drawbacks. For example, I could instead match using the rule title but this value isn't guaranteed to be unique and furthermore if the alert title is changed it should (correctly but silently) break the association. IMO, this behaviour is not intuitive for a migrated alert (especially for someone migrated to unified alerting for the first time). If you have thoughts here, I'm all ears!

The one-to-one mapping of notification policy to alert rule using a random `rule_uid` adds an extra layer of confusion to users performing the alerting migration (ie, what exactly is the `rule_uid` label? do newly-added rules need a similar `rule_uid` label? Why is the UI redundantly displaying the `rule_uid` redundant label when it's already the internal identifier?). It also leaves the migrated setup in a rather unmaintainable state. (For example, deleting a migrated alert rule doesn't delete the corresponding notification policy - #56053)

One alternate solution would be to attach `contact_point_[id|slug]=true` labels to migrated alerts, linking an alert rule to any number of contact points, and one migrated notification policy per contact point matching against the label and linking the contact point. (This would also provide a place to more cleanly migrate the 'Send Reminders' settings from a legacy notification channel to a 'Repeat interval' setting on the migrated notification policy; see #56578.)

Whatever the ideal solution, this issue represents a significant pain point in the current migration process. If the current process is in fact the best we can do, I'd like to see additional documentation in the upgrade guide explaining these additional generated resources and how one might reasonably maintain them moving forward, since it currently leaves a confusing mess of notification policies with no clear guidance on how to proceed.

### PR DESCRIPTION
**Pull request by JacobsonMT**
Previously, migrated legacy alerts were connected to notification policies through a `rule_uid` label in a 1:1 fashion. While this correctly mimicked pre-migration routing, it didn't create a notification policy structure that is easy to view/modify. In addition,
having one policy per migrated alert is, in some ways, counter to the recommended approach of Unified Alerting.

This change replaces `rule_uid`-based migrated notification policies with a private label called `__contacts__`. This label stores a list of double quoted strings containing the names of all contact points an AlertRule should route to (based on legacy notification channels). Then, one notification policy is created per contact point matching AlertRules via regex on this label.

**NOTE**: In order to simplify the created regex, we sanitize any legacy notification channel names that contain a double quote, replacing it with an underscore.

The result is a simpler, clearer, and easier to modify notification policy structure, with the added benefit that you can see which contact points an AlertRule is being routed to from the AlertRule creation page.

Fixes https://github.com/grafana/grafana/issues/56582

#### Legacy Setup

| Alert | Channels |
| ------------- | ------------- |
| Alert1 | email1, webhook |
| Alert2 | email2, webhook |
| Alert3 | webhook |
| Alert4 | email2 |
| Alert5 | email2 |

#### Previous Migrated Setup
5 Notification Policies: 
- One per alert, matched based on the alert's `rule_uid`
- Two (Alert1, Alert2) will have nested policies so that they send to multiple contact points.

Screenshot from 2022-07-11 18-22-37

#### New Migrated Setup
3 Notification Policies: 
- One per contact point, matched based on the `__contacts__` label regex, ex: `__contacts__=~.*"webhook".*`
- No nested policies.
- Policies match a collection of alerts instead of a single alert.
- Can add/remove which contact points are sent to on the alert rule page without changing policies by adding to the `__contacts__` label.
- Easy to see the purpose of each policy

Screenshot from 2022-10-17 20-12-55

Alert1 Labels:
Screenshot from 2022-10-17 20-13-57

### CODE REVIEW

**Comment by joeblubaugh** [comment]
I like this idea. How do we set the "consumption" policy in this case? ("Continue matching subsequent sibling nodes")? I *think* they all would need to be set with that to "true" for this approach to work.

**Comment by JacobsonMT** [comment]
> I like this idea. How do we set the "consumption" policy in this case? ("Continue matching subsequent sibling nodes")? I _think_ they all would need to be set with that to "true" for this approach to work.

Correct, each route is set to `continue=true` so that all siblings can match independently. See https://github.com/grafana/grafana/pull/52071/files#diff-3d6f33c060b6bf9fcfcb6c64e52c81198f5793d5ee389dd81395633cd8ea50f8R278.

**Comment by JacobsonMT** [comment]
> An interesting question came up, that I want to relay here. Should we have a single label on the alertRule (`__contacts__`, `migrated_contacts`, ... = `["webhook","email1"]`) that contains all contacts and is matched using regex or should we have multiple labels (`webhook=true`, `email1=true`) that match without needing regex?

Alernative would look something like this (given example setup in description):

Screenshot from 2022-07-15 19-39-06

Screenshot from 2022-07-15 19-39-50

**Comment by yuri-tceretian** [comment]
I like the idea. Do we control the name of the contact points? If yes, I think the single label with regex in notification policies would be a good solution.

**Comment by JacobsonMT** [comment]
> I like the idea. Do we control the name of the contact points? If yes, I think the single label with regex in notification policies would be a good solution.

Yes, we control the contact point names. They are migrated as is from the legacy notification channel names which are unique (https://github.com/grafana/grafana/blob/643d2bc890301526de36518a3c2c8c3ab0d718bb/pkg/services/alerting/store_notification.go#L308).

**Review by yuri-tceretian — APPROVED** [review]
LGTM! 👍🏻

File: pkg/services/sqlstore/migrations/ualert/channel.go [hunk 1]
```diff
@@ -115,19 +114,20 @@ func (m *migration) setupAlertmanagerConfigs(rulesPerOrg map[int64]map[*alertRul
...
-	if err != nil {
-		return "", err
-	}
+	return strings.Join(keys, ",")
+}
 
-	return string(jsonStr), nil
+// quote will surround the given string in double quotes and escape any existing interior double quotes.
+func quote(s string) string {
+	return `"` + strings.ReplaceAll(s, `"`, `\"`) + `"`
```
Comment (grobinson-grafana) [inline pkg/services/sqlstore/migrations/ualert/channel.go:130]: You might need to escape existing escaped quotes?

Comment (grobinson-grafana) [reply pkg/services/sqlstore/migrations/ualert/channel.go:130]: I'm not 100% sure on this, but if we have `\"` then we will get `\\"` when what we want I think is `\\\"`.

Comment (JacobsonMT) [reply pkg/services/sqlstore/migrations/ualert/channel.go:130]: You're right but extra escaping won't help either. Now that you point this out, I realize that my original intention for escaping quotes here doesn't accomplish what I had hoped.

Basically, I was attempting to use the outside quotes to separate contact names to make sure that partial names can't accidentally be matched by the policy regex `.*"<name>".*`. So, I escaped interior quotes so they wouldn't match the boundary. This doesn't work though, for example we could have two (rather evil) contact points:

- `hello\`
- `hello"world`

The regex for `hello\` would be `.*"hello\\".*` which would also incorrectly match `hello"world`'s contact label which would be `"hello\"world"`. [regex101 demonstration](https://regex101.com/r/3xOxye/1).

I'm sure there's a way to make the regex more robust using some lookahead or alternation but it won't be pretty... might be "good enough" in this state. Alternatively we can go the non-regex route and create multiple labels on each alert rule (one per contact).

Comment (JacobsonMT) [reply pkg/services/sqlstore/migrations/ualert/channel.go:130]: I think that the regex method is too brittle and too much hassle for the benefit. I'm going to push a change that uses multiple labels shortly and edit the description/screenshots.

Comment (JacobsonMT) [reply pkg/services/sqlstore/migrations/ualert/channel.go]: After some internal discussion, we decided to go back to the regex approach but sanitize any double quotes internal to the contact name on migration and replace them with `_`

File: pkg/services/sqlstore/migrations/ualert/channel.go [hunk 2]
```diff
@@ -114,7 +115,7 @@ func (m *migration) setupAlertmanagerConfigs(rulesPerOrg map[int64]map[*alertRul
 	return amConfigPerOrg, nil
 }
 
-// contactListToString creates a sorted string representation of a given map (set) of receiver names. Each name will be comma-separated and double-quoted. Double quotes inside a name are escaped.
+// contactListToString creates a sorted string representation of a given map (set) of receiver names. Each name will be comma-separated and double-quoted. Names should not contain double quotes.
 func contactListToString(m map[string]interface{}) string {
```
Comment (joeblubaugh) [inline pkg/services/sqlstore/migrations/ualert/channel.go:119]: Why is this restriction being relaxed?

(for my edification, not questioning the reason for the change)

**Review by joeblubaugh — APPROVED** [review]

File: pkg/services/sqlstore/migrations/ualert/channel.go (hunk 2 above)
Comment (joeblubaugh) [reply pkg/services/sqlstore/migrations/ualert/channel.go:119]: (I *think* it's because we already replace " with _)

**Pull request merged**